### PR TITLE
Adds GraphQL Schema Diff and Linting

### PR DESCRIPTION
**Pull request by redreceipt**
## DESCRIPTION

Screen Shot 2019-05-31 at 10 21 02 PM

Screen Shot 2019-05-31 at 10 50 07 PM

### What does this PR do, or why is it needed?

This PR does two things:
1. Gives us the ability to examine when the schema is updated and test for breaking changes via `yarn` and Travis
2. Enables client-side query lint checking against the schema.

### What design trade-offs/decisions were made?

Added a couple scripts and config files. Also adds a new stage in Travis. Added four new packages:
- `graphql-schema-linter`
- `get-graphql-schema`
- `@entria/graphql-findbreakingchanges`
- `eslint-plugin-graphql`

The first three could all be replaced with Prisma's `graphql-cli` except you invoke it with `graphql` and we already have a package called `graphql` so it throws an error 😞 

### How do I test this PR?

1. Make a breaking change to the schema (ex. change a field name)
2. `yarn schema-diff`
3. Write some bad code to a query in the app and make sure `eslint` yells at you.

### What automated tests did you write?

None.

### Possible Next Steps

1. **Fix schema formatting**. I added a `yarn schema-lint` command as well, very insightful but right now there are 328 errors, mostly just missing descriptions but possibly a few good ones as well. Maybe down the road, we can fix these and include the command in `yarn lint` for CI.
2. **Enable tracing**. This PR makes it impossible to commit breaking changes so we'll have to mark fields deprecated to slowly encourage client changes. When we do get ready to remove fields, we'll need a way to measure impact. This can be done by enabling tracing and acting on the recorded response data. 

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [x] Closes [#744]
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer
- [ ] fix current lint errors
- [x] update `schema.graphql` dynamically

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by redreceipt** [comment]
@vinnyjth only problem I'm not quite sure what to do about is how and when to update the local schema file. It has to be while `localhost:4000` is running. I could use the `get-introspection-data` script but I think the functionality should be contained to the `apollos-church-api` package. It should be able to update its own file somehow since the `yarn schema-diff` script is going to need it to function properly.

**Comment by redreceipt** [comment]
also I think I totally just solved those weird analytics errors we've been seeing with this linter haha

Screen Shot 2019-05-31 at 11 25 17 PM

**Comment by vinnyjth** [comment]
> also I think I totally just solved those weird analytics errors we've been seeing with this linter haha
> 
> Screen Shot 2019-05-31 at 11 25 17 PM

I wish :/ Those are client queries and so the linter isn't able to find them inside the `.graphql` file we generate. We will need to eslint disable those lines.

**Comment by redreceipt** [comment]
Is a "client query" just a query that pulls from a different GraphQL server? If that's the case, I think we can use multiple schemas so we could lint any query we have.

**Comment by vinnyjth** [comment]
> Is a "client query" just a query that pulls from a different GraphQL server? If that's the case, I think we can use multiple schemas so we could lint any query we have.

A client query is a query that resolves on the device itself, it doesn't hit any server, but is resolved from the on device cache. There isn't really a schema for the client side query, but I'm curious if you can find a way to lint from it.

**Comment by vinnyjth** [comment]
@redreceipt I can review this as soon as you are ready.

**Comment by redreceipt** [comment]
Yea, almost. I need a solution for the self updating `schema.graphql` file. Also would love your thoughts on the possible next steps I've outlined above.

**Comment by vinnyjth** [comment]
@redreceipt I think both of your next steps sound good, but are next steps for future issues. We can create PRs after we merge this.

**Comment by redreceipt** [comment]
🔥 🔥 🔥 love that! I'm gonna make a slight mod so it will automatically happen every `yarn start` using a `--start-server` param to the script that we'll use in Travis but won't have to use locally.

File: .travis.yml [hunk 1]
```diff
@@ -33,7 +33,7 @@ jobs:
         - travis_wait yarn codecov
     - stage: Schema Diff
       script:
-        - cd packages/apollos-chuch-api
+        - cd packages/apollos-church-api
```
Comment (redreceipt) [inline .travis.yml]: 😂

**Pull request merged**